### PR TITLE
MS-AMP support (w/o FSDP)

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Fixes up the MS-AMP integration to include:
- DeepSpeed support (1-2)
- Integration suite
- misaligned init

FSDP is **unsupported** while https://github.com/Azure/MS-AMP/pull/181 is in limbo. 

Closes https://github.com/huggingface/accelerate/issues/2143
Closes https://github.com/huggingface/accelerate/issues/2971
Closes https://github.com/huggingface/accelerate/issues/3050

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@BenjaminBossan @SunMarc

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -1388,17 +1395,23 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
 
         if self.native_amp:
             model._original_forward = model.forward
-            model_forward_func = model.forward.__func__ if hasattr(model.forward, "__func__") else model.forward
             autocast_context = get_mixed_precision_context_manager(self.native_amp, self.autocast_handler)
-            new_forward = autocast_context(model_forward_func)
-            if hasattr(model.forward, "__func__"):
-                model.forward = MethodType(new_forward, model)
-                model.forward = MethodType(convert_outputs_to_fp32(model.forward.__func__), model)
+            # NOTE: MS-AMP is special, and adds a `__func__` already to `model.forward`
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:1399]: Good idea to add a comment here, but at least for me, it's not obvious what is happening here even with the comment.

Also, it seems to be that the code path for `self.fp8_backend == "MSAMP"` and the code path for `not hasattr(model.forward, "__func__")` are the same, is that right? If so, would it make sense to combine them into a single condition?

```python
if (self.fp8_backend == "MSAMP") or not hasattr(model.forward, "__func__")`:
    ...
else:
    ...
```

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -1989,26 +2009,31 @@ def _prepare_ipex_or_xpu(self, *args):
...
 
-    def _prepare_msamp(self, *args):
+    def _prepare_msamp(self, *args, device_placement):
         if not is_msamp_available():
             raise ImportError(
                 "MS-AMP was not found on your system. Please ensure that MS-AMP is available "
                 " or choose `'te'` as the backend for FP8 mixed precision training."
             )
-        else:
+            # We've already checked for FSDP + MS-AMP during `__init__`
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:2018]: Wait, by removing the `else`, `msp` is only imported `if not is_msamp_available()`?

File: src/accelerate/accelerator.py [hunk 3]
```diff
@@ -2023,7 +2048,10 @@ def _prepare_msamp(self, *args):
                 result[i] = model
             elif isinstance(result[i], (torch.optim.Optimizer)):
                 result[i] = optimizer
-        return tuple(result)
+        if optimizer_index is not None:
+            # NOTE: MS-AMP moves the optimizer, but *not* the model to the right device
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:2052]: uff

File: src/accelerate/accelerator.py [hunk 4]
```diff
@@ -2116,7 +2144,9 @@ def prepare_optimizer(self, optimizer: torch.optim.Optimizer, device_placement=N
             return optimizer
         if device_placement is None:
             device_placement = self.device_placement
-        optimizer = AcceleratedOptimizer(optimizer, device_placement=device_placement, scaler=self.scaler)
+        # NOTE: Special case with MS-AMP we do *not* pass in the scaler, optimizer handles it for us
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:2147]: Above, it was stated that

> MS-AMP requires `GradScaler` even with bf16 autocast w/ single GPU or DDP:

Probably not a contradiction, but I found it confusing.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for addressing these issues. Overall, this looks good, I have a couple of comments but no big ones.

Nice benchmarks too, I didn't check all the details, as they're based on the existing ones AFAICT.

File: src/accelerate/accelerator.py (hunk 4 above)
Comment (muellerzr) [reply src/accelerate/accelerator.py]: Made it a bit more clear

**Review by SunMarc — APPROVED** [review]
Thanks for adding the support with Deepspeed ! Nice benchmarks for fp8. Just a small nit

**Pull request merged**